libraries/chain/controller.cpp

```diff
@@ -1519,48 +1544,52 @@ struct controller_impl {
               ("current_head_id", head->id)("current_head_num", head->block_num)("new_head_id", new_head->id)("new_head_num", new_head->block_num) );
          auto branches = fork_db.fetch_branch_from( new_head->id, head->id );
 
-         for( auto itr = branches.second.begin(); itr != branches.second.end(); ++itr ) {
-            pop_block();
-         }
-         EOS_ASSERT( self.head_block_id() == branches.second.back()->header.previous, fork_database_exception,
-                     "loss of sync between fork_db and chainbase during fork switch" ); // _should_ never fail
-
-         for( auto ritr = branches.first.rbegin(); ritr != branches.first.rend(); ++ritr ) {
-            optional<fc::exception> except;
-            try {
-               apply_block( (*ritr)->block, (*ritr)->is_valid() ? controller::block_status::validated : controller::block_status::complete );
-               fork_db.mark_valid( *ritr );
-               head = *ritr;
-            } catch (const fc::exception& e) {
-               except = e;
+         if (branches.second.size()) { // cater a case of switching from fork pending head to lib
```

Comment:
This is solely for the startup case where the node is switched from irreversible mode to speculative mode that we need to call ```maybe_switch_forks( fork_db.pending_head(), controller::block_status::validated );``` in order to let the state moves from LIB to pending_head. In this case branches.second() is null so nothing to roll back.

Reply:
Yes, but I would expect that startup case to be automatically taken care of by the `replay( shutdown ); // replay reversible blocks` line(s) that get called in  the `init` function. Now, the reason it is not is because of another bug where reversible blocks are not being written out in irreversible mode (see https://github.com/EOSIO/eos/pull/6842#discussion_r).

Reply:
Basically, I think the proper solution should result in maybe_switch_forks never being called when nodeos is in irreversible mode.